Data/HashMap/Base.hs

```diff
@@ -698,120 +731,111 @@ insert' h0 k0 v0 m0 = go h0 k0 v0 0 m0
-            !st' = go h k x (s+bitsPerSubkey) st
+            !st' = go h s k x (bs+bitsPerSubkey) st
         in Full (update16 ary i st')
-      where i = index h s
-    go h k x s t@(Collision hy v)
-        | h == hy   = Collision h (snocNewLeaf (L k x) v)
+      where i = index h bs
+    go h s k x bs t@(Collision hx l1 l2 hmx)
+        | h == hx   = Collision hx l1 l2 $ go (hashWithSalt (nextSalt s) k) (nextSalt s) k x 0 hmx
         | otherwise =
-            go h k x s $ BitmapIndexed (mask hy s) (A.singleton t)
-      where
-        snocNewLeaf :: Leaf k v -> A.Array (Leaf k v) -> A.Array (Leaf k v)
-        snocNewLeaf leaf ary = A.run $ do
-          let n = A.length ary
-          mary <- A.new_ (n + 1)
-          A.copy ary 0 mary 0 n
-          A.write mary n leaf
-          return mary
+            go h s k x bs $ BitmapIndexed (mask hx bs) (A.singleton t)
 {-# NOINLINE insertNewKey #-}
 
 
 -- Insert optimized for the case when we know the key is in the map.
 --
--- It is only valid to call this when the key exists in the map and you know the
--- hash collision position if there was one. This information can be obtained
--- from 'lookupRecordCollision'. If there is no collision pass (-1) as collPos
--- (first argument).
---
 -- We can skip the key equality check on a Leaf because we know the leaf must be
--- for this key.
-insertKeyExists :: Int -> Hash -> k -> v -> HashMap k v -> HashMap k v
-insertKeyExists !collPos0 !h0 !k0 x0 !m0 = go collPos0 h0 k0 x0 0 m0
+-- for this key. We still need to do the equality check in case of collisions.
+insertKeyExists :: (Eq k, Hashable k) => Hash -> Salt -> k -> v -> HashMap k v -> HashMap k v
+insertKeyExists !h0 !s0 !k0 x0 !m0 = go h0 s0 k0 x0 0 m0
   where
-    go !_collPos !h !k x !_s (Leaf _hy _kx)
+    go !h _ !k x !_s (Leaf _hy _kx)
         = Leaf h (L k x)
-    go collPos h k x s (BitmapIndexed b ary)
+    go h s k x bs (BitmapIndexed b ary)
         | b .&. m == 0 =
             let !ary' = A.insert ary i $ Leaf h (L k x)
             in bitmapIndexedOrFull (b .|. m) ary'
         | otherwise =
             let !st  = A.index ary i
-                !st' = go collPos h k x (s+bitsPerSubkey) st
+                !st' = go h s k x (bs+bitsPerSubkey) st
             in BitmapIndexed b (A.update ary i st')
-      where m = mask h s
+      where m = mask h bs
             i = sparseIndex b m
-    go collPos h k x s (Full ary) =
+    go h s k x bs (Full ary) =
         let !st  = A.index ary i
-            !st' = go collPos h k x (s+bitsPerSubkey) st
+            !st' = go h s k x (bs+bitsPerSubkey) st
         in Full (update16 ary i st')
-      where i = index h s
-    go collPos h k x _s (Collision _hy v)
-        | collPos >= 0 = Collision h (setAtPosition collPos k x v)
-        | otherwise = Empty -- error "Internal error: go {collPos negative}"
+      where i = index h bs
+    go h s k x bs t@(Collision hx l1@(L k1 v1) l2@(L k2 v2) hmx)
+        | h == hx   =
```

Comment:
The `h == hx` test is guaranteed to succeed in this context because we know the key is in the map.